### DISCUSSION TITLE
HRE accounts vs. burner ones

### DISCUSSION THREAD

**Discussion by kavehtehrani** [scaffold-eth/scaffold-eth-2#1004] [Q&A] [answered]
When using scaffold eth 2 deployment locally, I understand that hardhat creates account and displays their private keys when running "yarn chain". I also understand that scaffold eth creates burner wallets in browser per https://www.quicknode.com/guides/ethereum-development/smart-contracts/introduction-to-scaffold-eth

> Upon opening the app, you'll encounter a wallet section that provides the option to use a Burner wallet for transactions or connect to other external wallets of your choice. It's important to note that the app utilizes the Burner wallet for user interactions, distinct from the accounts generated by the local Hardhat network, which are primarily used for deploying contracts. For development and testing on localhost, the app features a faucet function to request ETH, ensuring you have the necessary funds to interact with smart contracts on the local Hardhat network

Testing locally this has created some roadblocks for me because I take the private key I see for the first named account and put it in my metamask wallet with hardhat network configured as below:

Network Name: Localhost 8545
RPC URL: http://127.0.0.1:8545/
Chain ID: 31337
Currency Symbol: ETH

I am not sure exactly what goes wrong from here, because even though I see myself connected with the address displayed, after this every time I try to do anything I get an error "Internal JSON-RPC error." which to me means either there isn't enough gas or something is wrong about the network configuration. 

I used https://scaffold-eth-assistant.streamlit.app/ to search for an answer, this is likely the problem:

> Based on the provided context, the issue you're facing is likely due to a nonce mismatch. When you use the private key from the Hardhat account in your Metamask wallet, it may propose a wrong nonce, resulting in the "Internal JSON-RPC error".
> 
> To resolve this, you have two options:
> 
> Reset the nonce manually: Set the nonce to 0 (for the first transaction of your address in your Hardhat environment).
> Use a burner wallet: Forget about the nonce issue and use a burner wallet instead. Burner wallets are designed for local development and are a better fit for this use case.
> It's also important to note that the accounts generated by Hardhat are meant for deploying contracts, not for interacting with the contract as the owner. The burner wallet, on the other hand, is designed for user interactions.
> 
> By using a burner wallet, you can avoid the nonce mismatch issue and interact with your contract as the owner without any issues.
> 

My main goal is to be able to interact locally with the contract as the owner and am somehow unable to properly interact as the owner.

This explains the issue but doesn't quite solve it. How can I make sure the owner that deploys the contract is the same wallet (burner or not) to interact with it? Ideally the burner wallet would be the same that deploys it. Is resetting the nonce the only way to interact as the owner? Thanks and apologies if I am missing something obvious here.

**Comment by carletex** [comment, marked as answer]
Hey @kavehtehrani thanks for this! This will help us to make SE-2 better and more intuitive.

> I understand that hardhat creates account and displays their private keys when running "yarn chain"

Yes! When you are deploying locally, the `deployer` account in `00_deploy_your_contract.ts` is always the first (well the 0th :D) hardhat address (0xf39...). 

> I am not sure exactly what goes wrong from here, because even though I see myself connected with the address displayed, after this every time I try to do anything I get an error "Internal JSON-RPC error."

This is probably the nonce error that you mention later (metamask can't detect the correct nonce on the hardhat network). It's annoying, yes.

A SE-2 Burner wallet is generated when you load SE-2 for the first time. PK is stored in local storage (ofc this is not secure at all, and burner wallets are only meant to be used locally). The good thing is that you can go fast, and no need to click "sign/send" when broadcasting a tx.

Currently, the only way to make 0xf39 (hardhat 0th account) your burner wallet would be to open your browser dev tools, and manually tweak the `burnerWallet.pk` local storage with the hardhat pk. Reload the page, and your burner should be 0xf39. A bit hacky, yes :')

---

Some random ideas that we could implement:
- A menu item in the account dropdown ("Import burner PK") 
- A menu item to set 0xf39 / hardhat account
- Have an env var in the NextJS app (e.g. BURNER_PK)... and then SE-2 use it if present. 

Always tricky when playing with PK, even if they are burners!

cc @rin-st @technophile-04

**Reply by damianmarti** [reply]
> * A menu item in the account dropdown ("Import burner PK")

@carletex I like this feature! It makes it more obvious how you can set your burner wallet to another address without needing to go to the dev console...

We did something similar in the burner wallet app.

Maybe we can create an issue to implement this. I can work on it...

**Comment by kavehtehrani** [comment]
I discussed this with someone more knowledgeable than me,  and this is another solution he suggested:

> You could prepare a simple script to transfer ownership, and then after you deploy the project you would open your frontend, pick your burner wallet address, then you would call the script from the CLI passing that address as argument for using the deployer's key to sign a transaction for transferring ownership for that new address.

**Comment by carletex** [comment]
Oh, if you just want to transfer ownership you can do it directly in the deploy script (`00_deploy_your_contract.ts`):

example 1: In a Contract that implements the Ownable pattern

```ts
  await deploy("YourContractOwnable", {
    from: deployer,
    args: [],
  });

  const YOUR_BURNER_ADDRESS = "0x0000....."; // Your burner/frontend address
  const yourContract = await hre.ethers.getContract<Contract>("YourContractOwnable", deployer);
  await yourContract.transferOwnership(YOUR_BURNER_ADDRESS);

```

example 2: If you are using our example contract (just need to pass it in the constructor)

```ts
  const YOUR_BURNER_ADDRESS = "0x0000.....";
  await deploy("YourContract", {
    from: deployer,
    args: [YOUR_BURNER_ADDRESS],
  });
```

**Comment by rin-st** [comment]
> Is resetting the nonce the only way to interact as the owner?

Most of the time it works without issues. What can help with that error: 
- Metamask -> Options (three dots) -> settings -> advanced -> clear activity tab data
- just click to faucet button (top right corner of your se-2 app) several times, so new hardhat network blocks will be mined. In rare cases it helps

**Reply by ck256-2000** [reply]
This is how I've been solving this problem for a couple years.... works.

**Comment by kavehtehrani** [comment]
Amazing, thanks for the suggestions this sorted me out! 

Btw where is the best place to document this? I looked at https://scaffold-eth-2-docs.vercel.app/ probably could benefit from a FAQ or common issues section. Happy to do a PR but wanted to make sure I haven't missed anything obvious.
